liberty-maven-plugin/src/main/java/net/wasdev/wlp/maven/plugins/server/DevMojo.java

```diff
@@ -0,0 +1,889 @@
+                        serverTask.waitForUpdatedStringInLog(regexp, timeout, logFile, messageOccurrences);
+                    }
+
+                    log.info("Running integration tests...");
+                    try {
+                        runIntegrationTests();
+                        log.info("Integration tests finished.");
+                    } catch (MojoExecutionException e) {
+                        Throwable cause = e.getCause();
+                        if (cause != null && cause instanceof MojoFailureException) {
+                            log.debug(e);
+                            log.error("Integration tests failed: " + cause.getLocalizedMessage());
+                        } else {
+                            log.error("Failed to run integration tests", e);
+                        }
+                    }
+                }
+            }
+
+            // finally, start watching for hotkey presses if not already started
+            util.runHotkeyReaderThread(executor);
+        }
+
+        @Override
+        public void checkConfigFile(File configFile, File serverDir) {
+            try {
+                ServerFeature servUtil = new ServerFeature();
+                Set<String> features = servUtil.getServerFeatures(serverDir);
+                features.removeAll(existingFeatures);
+                if (!features.isEmpty()) {
+                    List<String> configFeatures = new ArrayList<String>(features);
+                    log.info("Configuration features have been added");
+                    runMojo("net.wasdev.wlp.maven.plugins:liberty-maven-plugin", "install-feature", serverName,
+                            configFeatures);
+                    this.existingFeatures.addAll(features);
+                }
+            } catch (Exception e) {
+                log.debug("Failed to read configuration file", e);
+            }
+        }
+
+        @Override
+        public boolean compile(File dir) {
+            try {
+                if (dir.equals(sourceDirectory)) {
+                    log.info("Running maven-compiler-plugin:compile");
+                    runMojo("org.apache.maven.plugins:maven-compiler-plugin", "compile", null, null);
+
+                    log.info("Running maven-compiler-plugin:resources");
+                    runMojo("org.apache.maven.plugins:maven-resources-plugin", "resources", null, null);
+                }
+                if (dir.equals(testSourceDirectory)) {
+                    log.info("Running maven-compiler-plugin:testCompile");
+                    runMojo("org.apache.maven.plugins:maven-compiler-plugin", "testCompile", null, null);
+                    log.info("Running maven-compiler-plugin:testResources");
+                    runMojo("org.apache.maven.plugins:maven-resources-plugin", "testResources", null, null);
+                }
+                return true;
+            } catch (MojoExecutionException e) {
+                log.error("Unable to compile", e);
```

Comment:
Same thing here, I think we should fail out of dev mode if we failed to compile.

Reply:
Similar to the above, `MojoExecutionException` could be thrown for a number of possible reasons, including if the mojo ran properly but compilation failed on the user's project.  We wouldn't want to quit dev mode in this case, because the user needs to be allowed to correct their error.